File: StayPilot/StayPilot.Web/src/app/core/models/market-area-stats.ts
// Price numbers per place, worked out on the server and read back.
// Mirrors MarketAreaStatsResponse / MarketAreaLeaderboardResponse on the API.

// Which grain a row measures. Sent as its name, matching the AreaLevel enum on the API.
export type AreaLevel = 'District' | 'Municipality' | 'Town';

export interface MarketAreaStatsResponse {
  level: AreaLevel;
  district: string;
  municipality: string;
  town: string;
  // The place written out for a human: "Albufeira (Faro)".
  displayName: string;
  // How many listings the median came from. Always show it next to the price.
  listingCount: number;
  medianPricePerM2: number;
  medianAreaM2: number;
  // Listings asking clearly under what the model thinks they are worth. Under-priced, not cheap.
  belowEstimateCount: number;
  // Renovation stock: needs work, or an energy certificate of D or worse.
  projectCount: number;
  // Of projectCount: flagged by the advert itself vs. caught only by a poor energy grade.
  projectByConditionCount: number;
  projectByEnergyCount: number;
  projectMedianPricePerM2: number | null;
  projectMedianAreaM2: number | null;
  // The middle half of the project prices — what says whether the discount is real.
  projectP25PricePerM2: number | null;
  projectP75PricePerM2: number | null;
  moveInCount: number;
  moveInMedianPricePerM2: number | null;
  moveInMedianAreaM2: number | null;
  moveInP25PricePerM2: number | null;
  moveInP75PricePerM2: number | null;
  // Neither a project nor clearly finished, so left out of both sides but counted.
  unclassifiedCount: number;
  // Move-in price minus project price, in € per m². Null when either side is missing.
  renovationDiscountPerM2: number | null;
  // Why that discount should or should not be believed. Null when there is none to judge.
  renovationEvidence: RenovationEvidence | null;
  calculatedAtUtc: string;
}

export type Confidence = 'Low' | 'Medium' | 'High';

// Why one place's renovation discount deserves to be trusted. Two medians always differ by
// something; this is what says whether that something is a finding.
export interface RenovationEvidence {
  confidence: Confidence;
  // How much the two middle halves overlap, against the narrower one. 0 = clean separation.
  spreadOverlapPercent: number;
  // Share of the place's listings that got a verdict at all, project or finished.
  classifiedSharePercent: number;
  // The one-line reason behind the verdict, written on the server so every screen agrees.
  reason: string;
}

export interface MarketAreaLeaderboardResponse {
  items: MarketAreaStatsResponse[];
  // Null while the stats have never been worked out.
  calculatedAtUtc: string | null;
}

// The API filters, nothing more. Ordering happens in the browser: one level is a few hundred
// rows at most, so they all come over at once and re-sorting costs no request.
export interface MarketAreaLeaderboardQuery {
  level: AreaLevel;
  // Places with fewer listings than this are left out.
  minListings: number;
  // Narrow to one distrito, and inside it one município. Empty means the whole country.
  district?: string;
  municipality?: string;
}

// --- What your money buys -----------------------------------------------------------

export type Typology = 'T0' | 'T1' | 'T2' | 'T3' | 'T4' | 'T5' | 'T6' | 'T7' | 'T8' | 'T9' | 'T10';

export interface MarketAreaBudgetItemResponse {
  level: AreaLevel;
  displayName: string;
  district: string;
  municipality: string;
  town: string;
  // The most rooms the budget reaches here, on what that typology usually costs.
  bestTypology: Typology;
  medianPrice: number;
  medianAreaM2: number;
  medianPricePerM2: number;
  typologyListingCount: number;
  listingCount: number;
  // True when the place is only in reach because the budget was stretched.
  needsStretch: boolean;
  // Every typology the budget reaches here, most rooms first — not only the biggest.
  affordableTypologies: MarketAreaBudgetTypology[];
}

// One typology a budget reaches, as an alternative to the headline answer.
export interface MarketAreaBudgetTypology {
  typology: Typology;
  medianPrice: number;
  medianAreaM2: number;
  medianPricePerM2: number;
  listingCount: number;
}

export interface MarketAreaBudgetResponse {
  budget: number;
  // The budget after the stretch — what a place had to come in under to appear at all.
  reach: number;
  items: MarketAreaBudgetItemResponse[];
  calculatedAtUtc: string | null;
}

export interface MarketAreaBudgetQuery {
  budget: number;
  level: AreaLevel;
  minListings: number;
  district?: string;
  municipality?: string;
  // Leave out places where the budget does not reach at least this many rooms.
  minTypology?: Typology;
  // How far past the budget it may stretch, as a percentage. 0 = strict.
  stretchPercent?: number;
}

// --- Neighbour gaps -----------------------------------------------------------------

// One half of a pair, with the place broken into its parts so the screen can say which grain
// it is showing instead of leaving a bracket to be guessed at.
export interface NeighbourGapPlace {
  level: AreaLevel;
  district: string;
  municipality: string;
  town: string;
  // The place on one line, for the anchor picker and anywhere a table cell would not fit.
  displayName: string;
  // The price the gap was worked out from: all stock, or one typology's stock when narrowed.
  medianPricePerM2: number;
  listingCount: number;
  // The place's median across all its stock, whatever the comparison ran on. The two together
  // are the finding: "T2s 30% apart, all stock 4% apart" says the gap is about the flats.
  allStockPricePerM2: number;
  allStockListingCount: number;
}

export interface NeighbourGapResponse {
  expensive: NeighbourGapPlace;
  cheaper: NeighbourGapPlace;
  // Between the middle points of each place's listings, not between real borders.
  distanceKm: number;
  gapPercent: number;
}

export interface MarketAreaNeighbourGapResponse {
  items: NeighbourGapResponse[];
  // The typology every pair was compared on. Null means all stock at once.
  comparedOn: Typology | null;
  calculatedAtUtc: string | null;
}

export interface MarketAreaNeighbourGapQuery {
  level: AreaLevel;
  minListings: number;
  maxDistanceKm: number;
  minGapPercent: number;
  district?: string;
  municipality?: string;
  // Compare like with like. Without it a 30% gap can be entirely explained by one place
  // selling villas while the other sells studios.
  typology?: Typology;
  // The fewest listings of that typology a place needs to be half of a pair.
  minTypologyListings?: number;
}

File: StayPilot/StayPilot.Domain/Entities/MarketAreaStats.cs

using StayPilot.Domain.Enums;

namespace StayPilot.Domain.Entities
{
    /// <summary>
    /// Price numbers for one place, worked out from all the listings we have there.
    ///
    /// One row per place per level, so the same listing counts three times: once into its
    /// town row, once into its municipality row, once into its district row.
    ///
    /// Nothing here is typed in by hand. The whole table is rebuilt each time the stats are
    /// recalculated, the same way <see cref="PremiumFeature"/> is.
    /// </summary>
    public class MarketAreaStats
    {
        /// <summary>
        /// Database Id for this row.
        /// </summary>
        public int Id { get; set; }

        /// <summary>
        /// Which level this row measures: a district, a municipality or a town.
        /// </summary>
        public AreaLevel Level { get; set; }

        /// <summary>
        /// District this row belongs to. Always filled.
        /// </summary>
        public string District { get; set; } = string.Empty;

        /// <summary>
        /// Municipality this row belongs to. Empty on a district row.
        /// </summary>
        public string Municipality { get; set; } = string.Empty;

        /// <summary>
        /// Town this row belongs to. Empty on a district or a municipality row.
        /// </summary>
        public string Town { get; set; } = string.Empty;

        /// <summary>
        /// How many listings the price below was worked out from.
        /// Read this before the price: a median taken from three listings is not a market.
        /// </summary>
        public int ListingCount { get; set; }

        /// <summary>
        /// Middle price for each square meter in this place.
        /// The middle value and not the average, so one very expensive villa cannot drag it up.
        /// </summary>
        public decimal MedianPricePerM2 { get; set; }

        /// <summary>
        /// Middle floor area in this place. Says what the stock here actually is: a town of
        /// studios and a town of villas can share a price for each square meter.
        /// </summary>
        public decimal MedianAreaM2 { get; set; }

        /// <summary>
        /// How many listings here are asking clearly less than the valuation model thinks they
        /// are worth. This is the "deals" number, and it means under-priced rather than cheap -
        /// a whole town being inexpensive is not a bargain, one flat below its own worth is.
        /// </summary>
        public int BelowEstimateCount { get; set; }

        /// <summary>
        /// Middle point of this place's listings, used to find which places border which.
        /// Null when none of the listings here carry coordinates.
        ///
        /// Worked out from the listings rather than from a real boundary, because we hold no
        /// polygons. Good enough to tell neighbours apart, not a substitute for a border.
        /// </summary>
        public decimal? CentroidLatitude { get; set; }

        /// <inheritdoc cref="CentroidLatitude"/>
        public decimal? CentroidLongitude { get; set; }

        /// <summary>
        /// How many listings here look like renovation projects, meaning the advert says it needs
        /// work or its energy certificate is D or worse. See the calculator for why the
        /// certificate is used: "needs renovation" alone was on 1.4% of stock, too few to measure.
        /// </summary>
        public int ProjectCount { get; set; }

        /// <summary>
        /// Of <see cref="ProjectCount"/>, how many were flagged by the advert itself saying it
        /// needs renovation. Kept apart from the energy-grade count because the two signals are
        /// not equally trustworthy: this one is whatever the agent felt like typing.
        /// </summary>
        public int ProjectByConditionCount { get; set; }

        /// <summary>
        /// Of <see cref="ProjectCount"/>, how many were flagged only by an energy certificate of
        /// D or worse. The more objective of the two signals, and roughly ten times as common.
        /// A place whose projects are all this kind is measuring "poorly insulated", which is
        /// related to "needs work" but is not the same thing - worth being able to see.
        /// </summary>
        public int ProjectByEnergyCount { get; set; }

        /// <summary>
        /// Middle price for each square meter of the project stock here.
        /// Null when there are too few projects to take a median from.
        /// </summary>
        public decimal? ProjectMedianPricePerM2 { get; set; }

        /// <summary>
        /// Middle floor area of the project stock. Without it the discount is a rate with nothing
        /// to multiply it by, and "€420/m² cheaper" never becomes a sum of money.
        /// </summary>
        public decimal? ProjectMedianAreaM2 { get; set; }

        /// <summary>
        /// The middle half of the project prices for each square meter: a quarter of them ask
        /// less than <see cref="ProjectP25PricePerM2"/>, a quarter more than
        /// <see cref="ProjectP75PricePerM2"/>.
        ///
        /// This is the number that decides whether the discount is real. Two medians always
        /// differ by something; if the project spread and the move-in spread sit on top of each
        /// other, that difference is noise wearing a decimal point.
        /// </summary>
        public decimal? ProjectP25PricePerM2 { get; set; }

        /// <inheritdoc cref="ProjectP25PricePerM2"/>
        public decimal? ProjectP75PricePerM2 { get; set; }

        /// <summary>How many listings here are ready to move into.</summary>
        public int MoveInCount { get; set; }

        /// <summary>
        /// Middle price for each square meter of the move-in-ready stock here, which is what the
        /// project stock is discounted against. Null when there is too little to compare with.
        /// </summary>
        public decimal? MoveInMedianPricePerM2 { get; set; }

        /// <inheritdoc cref="ProjectMedianAreaM2"/>
        public decimal? MoveInMedianAreaM2 { get; set; }

        /// <inheritdoc cref="ProjectP25PricePerM2"/>
        public decimal? MoveInP25PricePerM2 { get; set; }

        /// <inheritdoc cref="ProjectP25PricePerM2"/>
        public decimal? MoveInP75PricePerM2 { get; set; }

        /// <summary>
        /// Listings here that are neither a project nor clearly move-in ready - an unknown
        /// condition with no certificate to fall back on.
        ///
        /// They are counted and then left out of both sides, which is the honest thing to do and
        /// also the thing most worth showing: a place with 1,200 listings whose renovation
        /// discount rests on 40 projects and 300 finished homes has 860 listings with no opinion,
        /// and a reader who cannot see that will trust the discount more than it deserves.
        /// </summary>
        public int UnclassifiedCount { get; set; }

        /// <summary>
        /// One row per typology found here (T1, T2, T3...), for answering what a budget buys.
        /// </summary>
        public List<MarketAreaTypologyStats> TypologyStats { get; set; } = new();

        /// <summary>
        /// When these numbers were last worked out (UTC time).
        /// </summary>
        public DateTime CalculatedAtUtc { get; set; } = DateTime.UtcNow;
    }
}
